File: frontend/src/types/workflow.ts
export type WorkflowStatus = "available" | "not_available" | "loading" | "error" | "no_execution_history";

export interface WorkflowResponse<T> {
  status: WorkflowStatus;
  reason?: string;
  data?: T | null;
}

export interface WorkflowSummaryData {
  id: string;
  job_id?: string;
  candidate_id?: string;
  status: string;
  started_at?: string;
  completed_at?: string;
  current_node?: string;
  workflow_version: string;
}

export interface WorkflowGraphNode {
  id: string;
  type: string;
  data: Record<string, any>;
  position: { x: number; y: number };
}

export interface WorkflowGraphEdge {
  id: string;
  source: string;
  target: string;
  label?: string;
  type?: string;
}

export interface WorkflowGraphData {
  nodes: WorkflowGraphNode[];
  edges: WorkflowGraphEdge[];
  current_node_id?: string;
}

export interface WorkflowTimelineEntry {
  event_id: string;
  workflow_id: string;
  timestamp: string;
  state: string;
  node_id?: string;
  duration_ms?: number;
  metadata: Record<string, any>;
}

export interface WorkflowTimelineData {
  entries: WorkflowTimelineEntry[];
}

export interface WorkflowNodeExecution {
  node_id: string;
  status: string;
  start_time?: string;
  end_time?: string;
  duration_ms?: number;
  retries: number;
  inputs: Record<string, any>;
  outputs: Record<string, any>;
  errors: string[];
  warnings: string[];
  tool_invocations: number;
  memory_retrievals: number;
}

export interface WorkflowNodeHistoryData {
  executions: WorkflowNodeExecution[];
}

export interface WorkflowEvent {
  event_id: string;
  timestamp: string;
  category: string;
  severity: string;
  message: string;
  node_id?: string;
  metadata: Record<string, any>;
}

export interface WorkflowEventListData {
  events: WorkflowEvent[];
}

export interface WorkflowCheckpoint {
  checkpoint_id: string;
  created_at: string;
  resumed_at?: string;
  rollback_available: boolean;
  workflow_version: string;
  graph_version: string;
  prompt_version: string;
  metadata: Record<string, any>;
  state_snapshot: Record<string, any>;
}

export interface WorkflowCheckpointListData {
  checkpoints: WorkflowCheckpoint[];
}

export interface WorkflowStatisticsData {
  total_duration_ms: number;
  average_node_time_ms: number;
  retries: number;
  failures: number;
  human_approvals: number;
  checkpoint_count: number;
  tool_calls: number;
  memory_retrievals: number;
}

File: backend/src/application/schemas/workflow.py
from pydantic import BaseModel, Field
from typing import Optional, Any, Generic, TypeVar, List, Dict
from datetime import datetime
from uuid import UUID

T = TypeVar("T")

class WorkflowResponseDTO(BaseModel, Generic[T]):
    """Standard envelope for Workflow Platform API responses."""
    status: str  # "available", "not_available", "loading", "error", "no_execution_history"
    reason: Optional[str] = None
    data: Optional[T] = None

class WorkflowSummaryDTO(BaseModel):
    id: UUID
    job_id: Optional[UUID] = None
    candidate_id: Optional[UUID] = None
    status: str
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    current_node: Optional[str] = None
    workflow_version: str

# Graph Topology for React Flow
class WorkflowGraphNodeDTO(BaseModel):
    id: str
    type: str # e.g. 'customNode'
    data: Dict[str, Any]
    position: Dict[str, float]

class WorkflowGraphEdgeDTO(BaseModel):
    id: str
    source: str
    target: str
    label: Optional[str] = None
    type: Optional[str] = "smoothstep"

class WorkflowGraphDTO(BaseModel):
    nodes: List[WorkflowGraphNodeDTO]
    edges: List[WorkflowGraphEdgeDTO]
    current_node_id: Optional[str] = None

class WorkflowTimelineEntryDTO(BaseModel):
    event_id: UUID
    workflow_id: UUID
    timestamp: datetime
    state: str  # "Started", "Running", "Node Started", "Paused", "Human Approval", "Completed", etc.
    node_id: Optional[str] = None
    duration_ms: Optional[float] = None
    metadata: Dict[str, Any] = Field(default_factory=dict)

class WorkflowTimelineDTO(BaseModel):
    entries: List[WorkflowTimelineEntryDTO]

class WorkflowNodeExecutionDTO(BaseModel):
    node_id: str
    status: str
    start_time: Optional[datetime] = None
    end_time: Optional[datetime] = None
    duration_ms: Optional[float] = None
    retries: int = 0
    inputs: Dict[str, Any] = Field(default_factory=dict)
    outputs: Dict[str, Any] = Field(default_factory=dict)
    errors: List[str] = Field(default_factory=list)
    warnings: List[str] = Field(default_factory=list)
    tool_invocations: int = 0
    memory_retrievals: int = 0

class WorkflowNodeHistoryDTO(BaseModel):
    executions: List[WorkflowNodeExecutionDTO]

class WorkflowEventDTO(BaseModel):
    event_id: UUID
    timestamp: datetime
    category: str
    severity: str
    message: str
    node_id: Optional[str] = None
    metadata: Dict[str, Any] = Field(default_factory=dict)

class WorkflowEventListDTO(BaseModel):
    events: List[WorkflowEventDTO]

class WorkflowCheckpointDTO(BaseModel):
    checkpoint_id: str
    created_at: datetime
    resumed_at: Optional[datetime] = None
    rollback_available: bool = False
    workflow_version: str
    graph_version: str
    prompt_version: str
    metadata: Dict[str, Any] = Field(default_factory=dict)
    state_snapshot: Dict[str, Any] = Field(default_factory=dict)

class WorkflowCheckpointListDTO(BaseModel):
    checkpoints: List[WorkflowCheckpointDTO]

class WorkflowStatisticsDTO(BaseModel):
    total_duration_ms: float
    average_node_time_ms: float
    retries: int
    failures: int
    human_approvals: int
    checkpoint_count: int
    tool_calls: int
    memory_retrievals: int
